### DISCUSSION TITLE
Reading msquic logs

### DISCUSSION THREAD

**Discussion by ManickaP** [microsoft/msquic#1510] [Q&A] [answered]
I'm trying to debug incorrectly closed streams in msquic and I'm trying to make sense of the logs.
- I'm seeing stream addresses being reused, I assume there's pooling going on behind the scenes. Just to confirm, when I see: `[strm][0x7f46cc04bc20] Destroyed` it means the stream has been returned to the pool and this is the last message for that particular instance.
- `[ api] Enter 19 (0x7f46cc04bc20).` according to code **19** is stream close. This should be the last operation called on the stream.
- `[strm][0x7f46cc04bc20] Indicating QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE` must be observed happen before the close is called. Otherwise, the stream will end up in bad state.

**What is the expected behavior of msquic, when the stream is closed before `QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE` arrives?**

I understand this is an application error, which must be fixed. But I'm not observing any errors from msquic. What I'm also NOT observing is any new QUIC_CONNECTION_EVENT_PEER_STREAM_STARTED after the MAX_STREAMS has been reached. Thus, I'm left with hanging server side connection, waiting for new streams; and hanging HTTP/3 client waiting for the response. It all gets unblocked after a certain timeout, so it doesn't hang indefinitely.

**Comment by nibanks** [comment, marked as answer]
Bad things happen if you close without shutting down completely:
```c
_IRQL_requires_max_(PASSIVE_LEVEL)
void
QuicStreamClose(
    _In_ __drv_freesMem(Mem) QUIC_STREAM* Stream
    )
{
    if (!Stream->Flags.ShutdownComplete) {

        if (Stream->Flags.Started) {
            //
            // TODO - If the stream hasn't been aborted already, then this is a
            // fatal error for the connection. The QUIC transport cannot "just
            // pick an error" to shutdown the stream with. It must abort the
            // entire connection.
            //
            QuicTraceLogStreamWarning(
                CloseWithoutShutdown,
                Stream,
                "Closing handle without fully shutting down");
        }
```
You trigger an implicit abort, BUT we don't have an error code to use for the abort! So we eventually just close with `0`. 😢 Ideally, this should be a FATAL error and I kill the process, but I'm not sure I can practically do that...

**Reply by nibanks** [reply]
On the topic of a timeout unblocking it, what timeout? One at the QUIC layer or above?

**Reply by ManickaP** [reply]
Seems like it gets unblocked on this:
`[conn][0x7f46cc03c180] "IDLE" timer expired`

**Reply by nibanks** [reply]
So the connection eventually idles out and dies. I wouldn't call that "unblocked" really. So, on the side opposite of the one calling `StreamClose` before it should, what does it do when/if it receives a peer abort event with error 0? Is it still cleaning up appropriately? I'm still quite certain your problem is that you aren't completely shutting down on both sides.

**Reply by ManickaP** [reply]
Yes, I know we're not closing properly on the server side, I fixed that and it helped.

But before that, the client side (the one which closes properly) hadn't received any aborts. The stream got shutdown_complete and properly closed before the server (the faulty side) decided to prematurely close.

**Reply by nibanks** [reply]
Would you be able to collect a log for a minimal repro of that behavior? It is possible we have a bug in that scenario.

**Reply by ManickaP** [reply]
What I'm getting at here is that people will hit this, they will forget to properly read all the buffers, await shutdowns etc. And it seems that msquic is not letting us know that we made a mistake, it just results in a hang...

**Reply by ManickaP** [reply]
Sure, I have it, do you want just a single faulty call or the 100 streams with the hang? I have logs for both atm. Should I attach them here?

**Reply by nibanks** [reply]
Yes, please attach the small log with a brief description of what the app is doing and expects to happen please.

**Reply by ManickaP** [reply]
[quic_1stream.log](https://github.com/microsoft/msquic/files/6366704/quic_1stream.log)

This is a faulty scenario before the fix, single stream on a single connection.

On the server side stream is calling close before receiving `QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE`:
Line 1655: `[6][7ddd.7dff][01:47:31.590168][strm][0x7fde300166d0] Closing handle without fully shutting down`
Line 1660: `[6][7ddd.7dff][01:47:31.590172][strm][0x7fde300166d0] Indicating QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE`

On the client side, the corresponding stream has been already closed and cleaned up:
Line 1584: `[4][7ddd.7e03][01:47:31.566376][strm][0x7fddc0049a30] Indicating QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE`
Line 1589: `[1][7ddd.7e22][01:47:31.588584][ api] Enter 19 (0x7fddc0049a30).`
Line 1596: `[4][7ddd.7e03][01:47:31.588619][strm][0x7fddc0049a30] Destroyed`

The server side stream won't get destroyed, eventually leading to MAX_STREAM exhaustion, without any indication of such problem from msquic.

**Reply by nibanks** [reply]
Still looking at the logs, but I suspect you aren't draining all the received data:
```
1156 [4][7ddd.7e03][01:47:31.483804][C][TX][5]   STREAM ID:0 Offset:0 Len:50
1193 [1][7ddd.7dff][01:47:31.483937][strm][0x7fde300166d0] Recv State: 1
1247 [4][7ddd.7e03][01:47:31.508583][C][TX][6]   STREAM ID:0 Offset:50 Len:0 Fin
1308 [4][7ddd.7e03][01:47:31.508795][C][TX][5] ACKed (24.983 ms)
1312 [4][7ddd.7e03][01:47:31.508799][C][TX][6] ACKed (0.206 ms)
1351 [6][7ddd.7dff][01:47:31.516460][S][TX][7]   STREAM ID:0 Offset:0 Len:2
1408 [6][7ddd.7dff][01:47:31.516577][S][TX][8]   STREAM ID:0 Offset:2 Len:33
1474 [7][7ddd.7dff][01:47:31.516782][S][TX][7] ACKed (0.307 ms)
1479 [7][7ddd.7dff][01:47:31.516786][S][TX][8] ACKed (0.191 ms)
1529 [7][7ddd.7dff][01:47:31.563933][S][TX][9]   STREAM ID:0 Offset:35 Len:0 Fin
1629 [6][7ddd.7dff][01:47:31.589855][S][TX][9] ACKed (25.916 ms)
1655 [6][7ddd.7dff][01:47:31.590168][strm][0x7fde300166d0] Closing handle without fully shutting down
1659 [6][7ddd.7dff][01:47:31.590171][strm][0x7fde300166d0] Recv State: 3
```
The `[strm][0x7fde300166d0]` logs are from the server's perspective. As you can see the `Recv State` isn't going to `5` (`QUIC_STREAM_RECV_FIN`) after we receive `[C][TX][6]   STREAM ID:0 Offset:50 Len:0 Fin` from the client. Looking at the code, this means we're never getting this event:
```c
    QuicTraceLogStreamVerbose(
        ReceiveComplete,
        Stream,
        "Recv complete (%llu bytes)",
        BufferLength);
```
So you still have a pending receive somewhere.

All that being said, it is a bug that we are still getting stuck. I'm still looking into this part.

**Reply by nibanks** [reply]
I've opened https://github.com/microsoft/msquic/issues/1512 and created PR https://github.com/microsoft/msquic/pull/1513 to fix it.

**Reply by CarnaViire** [reply]
@nibanks could your PR possibly break again these scenarios? https://github.com/microsoft/msquic/issues/1384

**Reply by nibanks** [reply]
If you ever shutdown the receive direction while you have a receive pending and we've already received the FIN from the peer, this could happen. So yes, in the case where you have a receive outstanding, then we receive the FIN, which gets queued, and then you call shutdown, this would get stuck. This PR will fix all instances of this behavior.

**Reply by CarnaViire** [reply]
I see, in the scenarios in the issue #1384 the receiving party just reads patiently, neither aborting reads, not closing the stream. So I guess we'll be OK there and the SHUTDOWN_COMPLETE *will* wait until we've read everything, right?

**Reply by nibanks** [reply]
If you wait until SHUTDOWN_COMPLETE you will be fine. The problem only comes if you prematurely abort the receive path after you've already actually received the FIN, but just not delivered it to the app.

**Reply by ManickaP** [reply]
> So you still have a pending receive somewhere.

I was aware of that and that's been fixed in the app as well. It was manifesting as never receiving `SHUTDOWN_COMPLETE` and was easy to spot on the very first stream.
